packages/realm/src/schema/normalize.ts

```diff
@@ -387,16 +388,39 @@ function assertNotUsingShorthand(input: string | undefined, info: PropertyInfo):
 /**
  * Get an error message for an invalid property schema.
  */
-function errMessage({ objectName, propertyName }: PropertyInfo, message: string): string {
+function propSchemaErrorMessage({ objectName, propertyName }: PropertyInfo, message: string): string {
   return `Invalid type declaration for property '${propertyName}' on '${objectName}': ${message}`;
 }
 
+/**
+ * Generate an error caused by an invalid property schema.
+ * (Returning a function rather than the Error itself in order
+ * for the Error to only be created if needed.)
+ */
+function propError(info: PropertyInfo, message: string): () => SchemaParseError {
+  return () => new SchemaParseError(propSchemaErrorMessage(info, message));
+}
+
+/**
+ * Get an error message for an invalid object schema.
+ */
+function objectSchemaErrorMessage(objectName: string, message: string): string {
+  const displayName = objectName ? `object '${objectName}'` : "unnamed object";
+  return `Invalid schema for ${displayName}: ${message}`;
+}
+
+/**
+ * Generate an error caused by an invalid object schema.
+ */
+function objectError(objectName: string, message: string): () => SchemaParseError {
+  return () => new SchemaParseError(objectSchemaErrorMessage(objectName, message));
+}
+
 /**
  * Throw an error caused by an invalid property schema.
  */
-function error(info: PropertyInfo, message: string): never {
-  // TODO: Create a SchemaParseError that extends Error?
-  throw new Error(errMessage(info, message));
+function throwError(info: PropertyInfo, message: string): never {
+  throw new SchemaParseError(propSchemaErrorMessage(info, message));
```

Comment:
I find it a bit strange that we have a function which is just responsible for throwing 🤔